### ISSUE TITLE
e2e tests are flaky

### ISSUE THREAD

**Issue by nicksnyder** [sourcegraph/sourcegraph#2681]
![image](https://user-images.githubusercontent.com/754768/54163161-4e4a3980-4415-11e9-9f75-965d3c088747.png)

Failed:
- https://buildkite.com/sourcegraph/sourcegraph/builds/29205
- https://buildkite.com/sourcegraph/sourcegraph/builds/29204

Passed:
- https://buildkite.com/sourcegraph/sourcegraph/builds/29202
- https://buildkite.com/sourcegraph/sourcegraph/builds/29210

**Comment by ijsnow** [comment]
The tests both started failing at the same point in the test suite, but the error wasn't from the result of the test being unexpected. It seems like we're overwhelming puppeteer rather than writing flakey tests. The first test that fails comes from this error:

> Timeout - Async callback was not invoked within the 30000ms timeout specified by jest.setTimeout.

Then the rest are the following error:

> [2019-03-11T20:56:51Z] Protocol error (Page.navigate): Session closed. Most likely the page has been closed.

**Comment by chrismwendt** [comment]
This might have been caused by the ops incident where Sourcegraph.com was slow for ~1h https://sourcegraph.slack.com/archives/C08JA8Q1H/p1552338934058300

The e2e tests need to fetch extensions from the Sourcegraph.com registry that are enabled by default (e.g. sourcegraph/go on the mux.go file).

We could eliminate that dependency on Sourcegraph.com by uploading extensions to a private registry, but how would we get the JS bundle? We could build it from the latest master of each sourcegraph-* repository. I'm guessing this would take ~1 day to implement.

**Comment by ijsnow** [comment]
That sounds like a good idea. I was thinking it seemed like it could have been happening from something else, rather than puppeteer or flakey tests. And this explains why failures were only in extensions dependent code

**Comment by nicksnyder** [comment]
I agree that it would be great for sourcegraph/sourcegraph CI to not depend on sourcegraph.com

More generally, what if someone broke the basic code intel extension (which is its own repo) and published that to sourcegraph.com. Should that break CI on sourcegraph/sourcegraph? Ideally not, right?

**Comment by ijsnow** [comment]
We could continue with [sourcegraph/sourcegraph-debug-extension](https://github.com/sourcegraph/sourcegraph-debug-extension) and modify e2e tests to only test what really matters about sourcegraph.com:

- Displays a tooltip with the data that we expect
- Can change background color of a line
- Can add "after text" on lines

The tests do currently rely on external services to succeed. This would also simplify the tests.

**Comment by chrismwendt** [comment]
If we used Sourcegraph.com's latest extensions, we get e2e tests on them and can catch bugs faster, but it's not 100% deterministic (someone could update and break an extension out of band).

I prefer determinism in this case, so I added this to my 3.2 roadmap issue https://github.com/sourcegraph/sourcegraph/issues/2452

**Comment by nicksnyder** [comment]
I also prefer determinism in CI. Let's aim for that.

**Comment by chrismwendt** [comment]
Backlogging. Here's the work involved:

- Add steps to the e2e test setup to:
  - Disable the default language extensions
  - Generate an access token
  - Set up a private extension registry (might need a license)
- Clone, build, and publish the language extensions to the private registry using the access token

**Comment by nicksnyder** [comment]
Are the tests still flaky? If so, this shouldn't be backlogged. If not, then we need a better issue title for the remaining work.

**Comment by chrismwendt** [comment]
Currently, the tests will fail if Sourcegraph.com goes down, so they're flaky in the sense that it's possible for them to fail then pass on retry. It's not like the tests themselves are of poor quality, it's that any sourcegraph/server instance depends on Sourcegraph.com out of the box to fetch language extensions.

I'll rename the title to "E2e tests fail when Sourcegraph.com is down".
